USER_EXPERIENCE_DESIGN.md

```diff
@@ -0,0 +1,226 @@
+Even more surprising is a deviation in runtime behavior, like back pressure or
+error handling, since this often results in data loss.
+
+Examples:
+
+- Using the same `codec` option name in both sources and sinks that support it.
+- Defaulting to applying back pressure regardless of the component or topology.
+
+## Guidelines
+
+The following guidelines help to uphold the above principles.
+
+### Logical boundaries
+
+#### Edition boundaries
+
+Vector offers both an enterprise and open-source edition. Vector's internal
+state, AKA configuration, should not differ between the editions. An open-source
+user should be able to transition to the enterprise edition by simply adding
+a Datadog API key to their Vector configuration.
+
+#### Source & sink boundaries
+
+Vector sources and sinks should be plentiful and specific. When possible, they
+should align with a well-defined protocol. If a well-defined protocol is not
+available, or the use case does not allow it, then the source or sink should
+align with its upstream client or target service. Additionally, sources and
+sinks can overlap. It is preferable to offer specific sources and sinks composed
+with broad sources and sinks. This promotes discoverability, aligns with user
+intent, and builds confidence that Vector meets their specific use case.
+Ideally, Vector would offer hundreds of sources and sinks that specifically
+integrate with various systems. Maintainability should be achieved through
+composability.
+
+Examples:
+
+- A `syslog` source as opposed to a `syslogng` source since it aligns with the
+  Syslog protocol.
+- A `datadog_agent` source as opposed to a `datadog_api` source since it aligns
+  on intent and reduces scope.
+- Again, a `syslog` source _in addition_ to a `socket` source with the `codec`
+  option set to `syslog` since it is more specific and discoverable.
+
+#### Transform boundaries
+
+Opposite of source and sinks, Vector transforms should be broad and minimal.
+They should align with a high-level use case, like mapping, aggregator, or
+routing. The goal here is to reduce the number of choices a user makes and
+be opinionated with solutions. Our opinions should prioritize performance and
+reliability over ease of use (within reason), but we should strive to achieve
+both.
+
+Examples:
+
+- A `remap` transform as opposed to multiple `parse_json`, `parse_syslog`, etc
+  transforms.
+- A `filter` transform as opposed to a `filter_regex`, `filter_datadog_search`,
+  etc transforms.
+
+### Naming
+
+#### Option naming
+
+Configuration option names should adhere to the following rules:
+
+- Alphanumeric, lowercase, snake case format
+- Use nouns, not verbs, as names (e.g., `fingerprint` instead of `fingerprinting`)
+- Suffix options with their unit. (e.g., `_seconds`, `_bytes`, etc.)
+- Be consistent with units within the same scope. (e.g., don't mix second and milliseconds)
+- Don't repeat the name space in the option name (e.g., `fingerprint.fingerprint_bytes`)
```

Comment:
Maybe show the preferred example as well `fingerprint._bytes` instead of `fingerprint.fingerprint_bytes`

Reply:
Should it instead just be `fingerprint.bytes`, no `_`?